src/app/store/product-detail/product-detail.component.ts

```diff
@@ -0,0 +1,84 @@
+import { Component, OnInit, Input } from '@angular/core';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { ActivatedRoute, Params, Router } from '@angular/router';
+import { MatSnackBar } from '@angular/material';
+
+import { Product } from '../../core/store/product.model';
+import { StoreService } from '../../core/store/store.service';
+
+@Component({
+  selector: 'app-product-detail',
+  templateUrl: './product-detail.component.html',
+  styleUrls: ['./product-detail.component.scss']
+})
+export class ProductDetailComponent implements OnInit {
+  product: Product = new Product();
+  shoppingForm: FormGroup;
+  errors: any[] = [];
+  loading = false;
+
+  constructor(
+    private storeService: StoreService,
+    private formBuilder: FormBuilder,
+    private route: ActivatedRoute,
+    private router: Router,
+    private snackBar: MatSnackBar,
+  ) {
+  }
+
+  ngOnInit() {
+    this.initializeForm();
+    this.storeService.getProduct(this.route.snapshot.params['id']).subscribe(result => {
+      this.product = result;
+      if (this.product.sizes && this.product.sizes.length === 0) {
+        this.shoppingForm.get('size').clearValidators();
+        this.shoppingForm.get('size').updateValueAndValidity();
+      } else {
+        this.shoppingForm.get('size').setValidators(Validators.required);
+      }
+    })
+
+    this.route.params
```

Comment:
is the logic below necessary? it looks like ngOnInit is calling getProduct twice and updating the components product twice?

Reply:
Yeah not quite sure what's going on here. Double calls to getProduct

Reply:
The below logic also has nested observables which is not good. If the below logic is required (which i suspect it is since ngOnInit is only fired once and we need to listen to param changes), then it should be refactored to chain the observables instead of nest them. Nesting observables or promises semi defeats the purpose, because you're right back in "callback hell", where error handling is very difficult and logic is hard to follow.

Reply:
My original design had the service storing an array of all products so the getProduct method was synchronous, just returning one product from the array. Phil implemented the product list component with the component holding the list of products instead of the service so when  I added getProduct to his version of the service I made it do an async call to the api to get the single product. I should have chained the observables when I did that instead of nest them.
As Rob pointed out, the purpose of the two calls to getProduct is the first loads the correct product when the page is loaded, the second loads the product if the param changes while we are on the detail page, so if someone changes the param in the address bar. I am thinking I should be using a resolver here to make sure the product is retrieved from the api before the page loads so I will look at doing that.

Reply:
I got the resolver working so now it subscribes to the service to get the product data and the component subscribes to the data so if it changes the component updates.